core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala

```diff
@@ -470,4 +470,45 @@ class GroupCoordinatorAdapter(
       expireTimestamp = expireTimestamp
     )
   }
+
+  override def deleteOffsets(
+    context: RequestContext,
+    request: OffsetDeleteRequestData,
+    bufferSupplier: BufferSupplier
+  ): CompletableFuture[OffsetDeleteResponseData] = {
+    val future = new CompletableFuture[OffsetDeleteResponseData]()
+
+    val partitions = mutable.ArrayBuffer[TopicPartition]()
+    request.topics.forEach { topic =>
+      topic.partitions.forEach { partition =>
+        partitions += new TopicPartition(topic.name, partition.partitionIndex)
+      }
+    }
+
+    val (groupError, topicPartitionResults) = coordinator.handleDeleteOffsets(
+      request.groupId,
+      partitions,
+      RequestLocal(bufferSupplier)
+    )
+
+    if (groupError != Errors.NONE) {
+      future.completeExceptionally(groupError.exception)
+    } else {
+      val response = new OffsetDeleteResponseData()
+      topicPartitionResults.forKeyValue { (topicPartition, error) =>
+        var topic = response.topics.find(topicPartition.topic)
```

Comment:
TIL that topics are ImplicitLinkedHashMultiCollection.Elements. 😅 
I'm curious why this specific response gets this typing compared to similar request/response jsons. But good to know that this find operation works and is fairly efficient.

Reply:
That’s correct. The « map key » does that.

Reply:
It was part of the protocol from the start. It basically means that when we deserialize, it builds a collection where the elements can be looked up via hashing. That's how we can use "find" here. (Not a new change, but I haven't seen it used often.)

Reply:
Map key is defined in the json for the request and response objects. Because it is define the message generator code gives this typing.

Reply:
Yeah, that's right. In the schema definition, you can specify that you want a map like data structure instead of a list. This is done by specifying the "mapKey" in the spec.

As Justine said, this was done before this patch so I just kept it as-is.

Reply:
aha, that makes sense. thanks! i noticed the changes in OffsetDeleteResponse.java were a bit different than the other APIs as well

Reply:
It is almost the same. The main difference is that we don't need another Map to lookup topics by their name because the ImplicitLinkedHashMultiCollection gives us this.